src/main/java/org/opensearch/knn/index/codec/KNN80Codec/KNN80DocValuesProducer.java

```diff
@@ -0,0 +1,154 @@
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.index.codec.KNN80Codec;
+
+import lombok.NonNull;
+import lombok.extern.log4j.Log4j2;
+import org.apache.lucene.codecs.DocValuesProducer;
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.NumericDocValues;
+import org.apache.lucene.index.SegmentReadState;
+import org.apache.lucene.index.SortedDocValues;
+import org.apache.lucene.index.SortedNumericDocValues;
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.FSDirectory;
+import org.apache.lucene.store.FilterDirectory;
+import org.opensearch.common.io.PathUtils;
+import org.opensearch.knn.common.KNNConstants;
+import org.opensearch.knn.index.codec.util.KNNCodecUtil;
+import org.opensearch.knn.index.engine.KNNEngine;
+import org.opensearch.knn.index.memory.NativeMemoryCacheManager;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.opensearch.knn.common.KNNConstants.MODEL_ID;
+import static org.opensearch.knn.index.mapper.KNNVectorFieldMapper.KNN_FIELD;
+
+@Log4j2
+public class KNN80DocValuesProducer extends DocValuesProducer {
+
+    private final SegmentReadState state;
+    private final DocValuesProducer delegate;
+    private final NativeMemoryCacheManager nativeMemoryCacheManager;
+    private final Map<String, String> indexPathMap = new HashMap();
+
+    public KNN80DocValuesProducer(DocValuesProducer delegate, SegmentReadState state) {
+        this.delegate = delegate;
+        this.state = state;
+        this.nativeMemoryCacheManager = NativeMemoryCacheManager.getInstance();
+
+        Directory directory = state.directory;
+        // directory would be CompoundDirectory, we need get directory firstly and then unwrap
+        if (state.directory instanceof KNN80CompoundDirectory) {
+            directory = ((KNN80CompoundDirectory) state.directory).getDir();
+        }
+
+        FSDirectory fsDirectory = null;
+        try {
+            fsDirectory = ((FSDirectory) FilterDirectory.unwrap(directory));
+        } catch (ClassCastException ex) {
+            log.warn("{} can not casting to FSDirectory", directory.toString());
+            return;
+        }
+        assert fsDirectory != null;
+        String directoryPath = fsDirectory.getDirectory().toString();
+        for (FieldInfo field : state.fieldInfos) {
+            if (!field.attributes().containsKey(KNN_FIELD)) {
+                continue;
+            }
+            // Only Native Engine put into indexPathMap
+            KNNEngine knnEngine = getNativeKNNEngine(field);
```

Comment:
getNativeKNNEngine  just want to verify the native Engine type, and get engine extension like `KNN80DocValuesConsumer` do, but we need ignore model's engine. so i think `FieldInfoExtractor` would not be hear

Reply:
> getNativeKNNEngine just want to verify the native Engine type

Then in that case the name of the function is misleading. More so you are not just validating the Native engines you are validating if they are faiss/Nmslib. Atleast the way I would recommend here to do is, use the FieldInfoExtractor to find the engine and then compare it againsts the list of engines you want this logic to run.

This will ensure that code is clearly telling what it is doing, also the logic of extracting KNNEngine is in 1 place. :)

Reply:
> use the FieldInfoExtractor to find the engine

i am a little bit confused, `FieldInfoExtractor` only enum the field type, when there is Lucene engine which is not native engine, that there is no need to added into `indexPathMap` which in `KNN80DocValuesProducer`, but `FieldInfoExtractor` always return `FLOAT` as the fact we need nativeEngine type.

and `getNativeKNNEngine` function java docs says `if and only if Native Engine we return specific engine, else return null` so we can get native engine type and use common logic `KNNCodecUtil.getEngineFiles`

Reply:
@luyuncheng I think you need to pull in new changes from main. There is a new function added in FieldInfoExtractor. Ref: https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/common/FieldInfoExtractor.java#L30-L36

> also getNativeKNNEngine function java docs says if and only if Native Engine we return specific engine, else return null so we can get native engine type and use common logic KNNCodecUtil.getEngineFiles

I understand your concern and yes `FieldInfoExtractor.extractKNNEngine` will give the engine if present or give the default engine. But in this case we want to do 2 things:
1. Get the engine if the engine is set.
2. If the engine is set then it should not be Lucene Engine.


[Suggestion] Here is something I think we can do:
1. `field.attributes().containsKey(KNNVectorFieldMapper.KNN_FIELD)` use this to know if the field is KNNType or not.
2. If the field is KNNType then get the engine out of FieldInfo.
3. Validate if engine == FAISS or NMSLIB.

I see 2 is already taken care by `FieldInfoExtractor.extractKNNEngine` so if we enhance `FieldInfoExtractor.extractKNNEngine` to add 1 we can reuse the code, and ensure we have a very tight implementation of extracting engine from FieldInfo. Rather than having logic to extract kNN engine at multiple places.

The 3 you can do in a private function in this class.

Let me know if this sounds good. My intension here is to ensure that this logic of getting KNNEngine is at one place and not duplicated.

Reply:
i like the idea! so in https://github.com/opensearch-project/k-NN/pull/1946/commits/07806261e34cd4d22fa98897fad4b60a65181d57 i try to use `FieldInfoExtractor.extractKNNEngine`